packages/datadog-instrumentations/src/helpers/hook.js

```diff
@@ -0,0 +1,41 @@
+'use strict'
+
+const path = require('path')
+const iitm = require('../../../dd-trace/src/iitm')
+const ritm = require('../../../dd-trace/src/ritm')
+
+function Hook (modules, onrequire) {
+  if (!(this instanceof Hook)) return new Hook(modules, onrequire)
+
+  this._patched = {}
+
+  const safeHook = (moduleExports, moduleName, moduleBaseDir) => {
+    const parts = [moduleBaseDir, moduleName].filter(v => v)
+    const filename = path.join(...parts)
+
+    if (this._patched[filename]) return moduleExports
+
+    this._patched[filename] = true
+
+    return onrequire(moduleExports, moduleName, moduleBaseDir)
+  }
+
+  this._ritmHook = ritm(modules, {}, safeHook)
+  this._iitmHook = iitm(modules, {}, (moduleExports, moduleName, moduleBaseDir) => {
+    // Default export is always moved to `default` for CommonJS modules.
+    if (moduleExports && moduleExports.default) {
+      moduleExports.default = safeHook(moduleExports.default, moduleName, moduleBaseDir)
```

Comment:
I'm not quite sure I understand what is going on here. Any ESM module with a default export would go through this?

Reply:
sure, maybe just a note so there is a breadcrumb if things do upgrade

Reply:
There is already a comment about this. What should be clarified?

Also, can we detect if a module is CommonJS vs ESM from the loader hook? This would allow handling this more cleanly and in a forward compatible way.

Reply:
@rochdev if you just add "which are the only type of modules instrumented by dd-trace-js right now." or something like that, then that should be sufficient explanation for now I think.